Time series

1Kabman1 opened the pull request:
Рен привет ))

1Kabman1 committed:
modified:   internal/services/handler.go

1Kabman1 committed:
Added two new  structures aTime abd aTimeDuration ,

1Kabman1 committed:
I fixed the code. There were various warinds

1Kabman1 committed:
Deleted flag from internal/services/services.go of prepareTheDataForHashing

1Kabman1 committed:
I created 1 new method IncreaseValue,  and transferred the logic to the package hashStorage

1Kabman1 committed:
Decision

1Kabman1 committed:
Added list

1Kabman1 committed:
I did the task, the time interval

1Kabman1 committed:
Added a correct calculation for total value

ribatulin commented on internal/hashStorage/time.go:
обычно файл  должен называться идентично описываемому в нем классу, тоесть timeDuration.go. Везде по аналогии надо, в едином стиле

1Kabman1 replied:
Сделано

ribatulin commented on internal/hashStorage/time.go:
```diff
@@ -0,0 +1,19 @@
+)
+
+type timeDuration struct {
+	Duration int
```
тут особо нигде не ясно, что это секунды. Либо в названии отразить, либо комментарий написать в свойству

1Kabman1 replied:
Принял исправл  на  DurationSec

1Kabman1 committed:
Change struct time.go

1Kabman1 committed:
Merge branch 'timeSeries' of github.com:1Kabman1/Antifraud-payment-system into timeSeries

1Kabman1 committed:
renamed time.go into timeDuration.go

ribatulin commented on README.md:
```diff
@@ -50,34 +51,40 @@ bashScript остановит программу. Так же он провер
+         "bank_card"
+      ],
+      "AggregateValue": "amount",
+      "Duration":{"Duration":39732}
```
мутный формат. В запросе duration это строка неясного, неописанного формата, а в ответе в том же ключе Duration приходит уже объект , в котором есть так же свойство Duration, но оно уже int в секундах. 

Не надо вносить путаницу - если задаешь формат строкой в ключе Duration, то и возвращать его надо аналогично, если подразумевается, что возвращается тот же обхект правила (а это именно так). 

В идеале я бы вообще сделал Duration всегда просто int , обозначающий всегда секунды.

всю эту конвертацию строк 02h40m15s в количество секунд может делать другая система. например веб страница, при отправке запроса  на создание правил. 

Давай сделаем везде Duration - это кол-во секунд, без разночтений и введений формата  виде строки, бещ вложенного объекта в ответе
Сервис тут должен быть топорный, и предоставлять простое и понятное API

1Kabman1 replied:
исправил , удалил вообще эту структуру

ribatulin commented on README.md:
```diff
@@ -103,7 +110,8 @@ localhost:
     ],
-    "AggregateValue": "amount"
+    "AggregateValue": "amount",
+     "Duration":{"Duration":39732}
```
Duration: int везде, какой смысл вкладывать Duration: {"Duration": int} - не ясно, не красиво и тавтология

1Kabman1 replied:
Удалил структуру Duration

ribatulin commented on internal/hashStorage/timeDuration.go:
```diff
@@ -0,0 +1,19 @@
+package hashStorage
+
+import (
```
от этого надо избавиться сейчас, просто сделав Duration: int свойством правила, без отдельного объекта

1Kabman1 replied:
Сделано

ribatulin commented on internal/hashStorage/order.go:
```diff
@@ -0,0 +1,11 @@
+package hashStorage
+
+type Order struct {
```
почему названо Order - это заказ? или порядок?

ribatulin replied:
TimeSeriesColumnValue давай. "Значение колонки временных рядов."

1Kabman1 replied:
Выполнено

ribatulin commented on internal/hashStorage/rule.go:
```diff
@@ -2,9 +2,10 @@ package hashStorage
+	Name              string       `json:"Name"`
+	AggregateBy       []string     `json:"AggregateBy"`
+	AggregateValue    string       `json:"AggregateValue"`
+	Duration          timeDuration `json:"DurationSec"`
```
просто int, тут всегда секунды

1Kabman1 replied:
```
type Rule struct {
	AggregationRuleId int
	Name              string   `json:"Name"`
	AggregateBy       []string `json:"AggregateBy"`
	AggregateValue    string   `json:"AggregateValue"`
	ExpirationTime    int      `json:"ExpirationTime"`
	TimePeriod        int      `json:"TimePeriod"`
}
```
Теперь так только int

ribatulin commented on internal/hashStorage/counter.go:
```diff
@@ -1,10 +1,30 @@
-	Value int
+	id         int
+	TotalValue int
+	Values     *list.List
```
TimeSeriesValues - временные ряды значений давай назовем

1Kabman1 replied:
Выполнено

ribatulin commented on internal/hashStorage/storage.go:
```diff
@@ -55,19 +61,40 @@ func (s *Storage) HasRule(id int) bool {
+func (s *Storage) IncreaseValue(key [16]byte, AggregateValue string,
+	aAmount float64, duration int) {
+	_, c := s.Counter(key)
+	ord := NewOrder()
```
зачем создават новое значение в timeSeries (массиве значений за едииницу времени) на каждое увеличение. 

получится тут будет каждый вызов  Storage.IncreaseValue внутри Counter в Values добавлять новый объект.
но зачем тогда мы собираем агрегаты, тут просто в память не поместится тогда все с таким подходом.

Смотри, TimeSeries данные это как бы колонки -субагрегатов за конкретное время с определенной точностью.

к примеру,  есть правило "сумма платежей по карте за 1 последнюю минуту с точностью 5 секунд."

значит наш Counter будет содержать в timeSeriesValues (приватное свойство типа List)

максимум 12 актуальных значений (1минута=60 секунд Duration / точность в 5 секунд ) 

с 0 секунды по 5 невключительно
с 5 секунды по 10 невключительно
с 10 секунды по 15 невключительно
......
......
с 55 секунды по 60 невключительно

и таким образом, когда в наш счетчик приходит новое значение, он всего лишь должен найти, в какой из столбцов TimeSeries ему надо приплюсовать то, что надо приплюсовать.

А значением счетчика всегда будет просто сумма актальных элементов из TimeSeriesValues на момент запроса 

чуешь разницу? не создаем Order на каждый платеж, а ищем хранилище, соотвествующее времени его поступлания и уже туда плюсуем.

1Kabman1 replied:
Выполнил

ribatulin commented on internal/hashStorage/storage.go:
```diff
@@ -55,19 +61,40 @@ func (s *Storage) HasRule(id int) bool {
+		c.TotalValue += int(aAmount)
+		ord.Value = int(aAmount)
+	}
+	ord.T.DurationSec = int(time.Now().Unix()) + duration
```
ты тут хотел расчитать время экспирации, тогда тебе надо так и назвать свойство тут внутри:

expiresAt - истекает в ....
или еще вариант
actualTill - актуально до ....

Но этим должен управлять сам Counter, а не ты извне, залезая в кишки Counter'у
подумай о переносе части логики из этого метода в Counter, тк за увеличение значений  отвечать может Counter, предоставляя публичный интерфейс в виже методов, не раскрывая наружу всем свойства. Негоже ведь человеку ходить с расстегнутыми карманами)))

1Kabman1 replied:
Сделано

ribatulin commented on internal/hashStorage/storage.go:
```diff
@@ -55,19 +61,40 @@ func (s *Storage) HasRule(id int) bool {
-	s.AddToArchivist(idRule, idCounter)
+	if !s.HasCounter(key) {
+		aNewCounter := NewCounter()
+		aNewCounter.Values = list.New()
```
а зачем счетчик свои свойства (Values) сделал публичным))? я думаю, что структура/объект счетчикка не должны раскрывать свое внутреннее устройство и 

0) Values сделать приватным
1) aNewCounter.Values = list.New() делаться в конструкторе самого Counter, а не через публичный доступ

не надо давать внешнему коду объекта лишних знаний  о внутреннем устройстве другого.

когда такое происходит, значит нормально не выделен интерфейс/архитекура/абстракция, и это повод пересмотреть код. 

Я вижу, что в счетчике Counter должен появиться метод RegisterValue(....с некими аргументами, но без лишних...)

который будет находить верный столбец в TImeSeries данных и увеличивать его на значение, которое пришло в аргументах. и вот этот вот Counter.RegisterValue можно спокойно вызывать из IncreaseValue отсюда, когда мы 1) нашли счетчик нужный 2) поняли тип правила и определились, какое значение будем плюсовать в какой счетчик 3) и уже сам Counter сделает внутри себя 

c.Values.PushBack(ord)
и подчистку неактуального тоже. не раскрывая то, как он внутри устроен

1Kabman1 replied:
```
type Counter struct {
	id               int
	timeSeriesValues [][]int
	expirationTime   int
	timer            int
	timePeriod       int
}
```
Нет ни  одного публтчного поля в структуре

ribatulin commented on internal/hashStorage/order.go:
```diff
@@ -0,0 +1,11 @@
+
+type Order struct {
+	Value int
+	T     timeDuration
```
это время, до которого обхект актуален внутри списка TimeSeriesValues. писал, что его так и надо назвать 

actualTill или expiresAt  -смотря как больше по душе. первое скорее более актуально.

1Kabman1 replied:
Order удалил за ненадобностью

ribatulin commented on internal/hashStorage/counter.go:
```diff
@@ -1,10 +1,30 @@
+func (c *Counter) DeleteExpiredOnes() {
+	for i := c.Values.Back(); i != nil; {
+		ord := i.Value.(Order)
+		if ord.T.DurationSec > 0 && ord.T.DurationSec < int(time.Now().Unix()) {
```
должно происходить независимо как то от добавления значений при обработке, а сейчас  вызывается только оттуда. пока не трогаем, оставим на потом отдельной задачей.

1Kabman1 replied:
Сейчас работает на независимо на отдельной горутине по  заданому временому отрезку

1Kabman1 replied:
Данный метод больше не требуется

ribatulin reviewed (changes requested):
пока посмотрел, давай править и обсуждать

1Kabman1 committed:
Added logs

1Kabman1 committed:
Wrote the plan of time series

1Kabman1 committed:
some changes

1Kabman1 committed:
I wrote it but haven not checked it yet

1Kabman1 committed:
New decision of time series

1Kabman1 committed:
Renemed Value of Counter into timeSeriesValues

1Kabman1 committed:
Some changes. This app works

1Kabman1 committed:
The absence of time accounting for aggregated data is taken into account. The tests are also transferred to another file for compactness

1Kabman1 committed:
Transferred to another file for compactness

1Kabman1 committed:
Renamed  internal/hashStorage/storage.go -> internal/hashStorage/hashStorage.go
 & internal/hashStorage/storage_test.go -> internal/hashStorage/hashStorage_test.go
